### ISSUE TITLE
generator in mnist_acgan example collapses to black images

### ISSUE THREAD

**Issue by wlamond** [fchollet/keras#4585]
As the title states, when I run the mnist_acgan example the discriminator loss quickly goes to zero and the generator loss increases over time. The generated images at the end of each epoch are black. I'm digging into this myself, but GANs aren't my area so I expect progress to be slow. Any suggestions on where to look? Is anyone able to run this example without any problems?

Link to the example: https://github.com/fchollet/keras/blob/master/examples/mnist_acgan.py

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by bobchennan** [comment]
I mentioned this problem before in the pull requests. The solution is easy. You can change the random seed and that works for me

**Comment by wlamond** [comment]
Thank you for the suggestion. That didn't work for my unfortunately. I see you're using a different version of CUDA and cuDNN than I am. I'm using CUDA version 7.0 and cuDNN 5.0.4. I'm due for an update, so I'll report back when I test this again.

**Comment by bobchennan** [comment]
I think I upgraded the version and changed the seed to 1331. Then it works fine.

**Comment by isaacgerg** [comment]
Is this fixed by #4482

**Comment by wlamond** [comment]
This is working after updating to CUDA 8.0. Thanks for the tips!

**Comment by sautin1** [comment]
Hello!

I have CUDA 8.0 and the latest versions of keras and tensorflow. However, the generator keeps creating black images. I tried changing the random seed and it didn't help.

System overview:
Windows 10,
Python 3.5.2 :: Anaconda 4.2.0 (64-bit),
tensorflow-gpu==0.12.1,
Keras==1.2.1

Any suggestions on how to make the example work?

**Comment by isaacgerg** [comment]
Last I checked, which was about 10 weeks ago, my PR to fix this issue was
rejected and the fix put in by @fchollet did not fix the issue.  I gave up
and moved to tensorflow.

**Comment by sautin1** [comment]
As I understand, you address to this PR: https://github.com/fchollet/keras/pull/4482
Are you sure it has something to do with the issue being discussed? The mnist_acgan example doesn't seem to even call the function, which was corrected in your PR.

> Last I checked, which was about 10 weeks ago, my PR to fix this issue was
rejected and the fix put in by @fchollet did not fix the issue.  I gave up
and moved to tensorflow.

**Comment by isaacgerg** [comment]
@sautin1   Yes, you are correct, this doesnt look to be the same issue.  This should be pretty easy to debug though.  You should remove layers in the generator 1 by 1 until they produce an output.  If you get to the a single layer and the image is still black, you can trace through the code in the debugger to find the offending line.

**Comment by s-macke** [comment]
The examples also collaped into black images for me. I changed the learning rate from 0.0002 to 0.00005 and it converges now.

**Comment by CanoeFZH** [comment]
To change random seed and learning rate to 0.00005 do not work for me.
this is my training loss info, seems that d loss is too small which are not good:
Testing for epoch 1:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 8.35 | 6.03            | 2.32 
generator (test)       | 8.40 | 6.09            | 2.31 
discriminator (train)  | 1.32 | 0.01            | 1.31 
discriminator (test)   | 1.20 | 0.00            | 1.20 
Epoch 2 of 50
599/600 [============================>.] - ETA: 0s  
Testing for epoch 2:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 9.38 | 7.07            | 2.31 
generator (test)       | 8.89 | 6.58            | 2.30 
discriminator (train)  | 1.21 | 0.00            | 1.21 
discriminator (test)   | 1.19 | 0.00            | 1.19 
Epoch 3 of 50
599/600 [============================>.] - ETA: 0s  
Testing for epoch 3:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 9.79 | 7.48            | 2.31 
generator (test)       | 9.30 | 6.99            | 2.30 
discriminator (train)  | 1.20 | 0.00            | 1.20 
discriminator (test)   | 1.18 | 0.00            | 1.18

**Comment by CanoeFZH** [comment]
599/600 [============================>.] - ETA: 0s   
Testing for epoch 1:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 8.35 | 6.03            | 2.32 
generator (test)       | 8.40 | 6.09            | 2.31 
discriminator (train)  | 1.32 | 0.01            | 1.31 
discriminator (test)   | 1.20 | 0.00            | 1.20 
Epoch 2 of 50
599/600 [============================>.] - ETA: 0s  
Testing for epoch 2:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 9.38 | 7.07            | 2.31 
generator (test)       | 8.89 | 6.58            | 2.30 
discriminator (train)  | 1.21 | 0.00            | 1.21 
discriminator (test)   | 1.19 | 0.00            | 1.19 
Epoch 3 of 50
599/600 [============================>.] - ETA: 0s  
Testing for epoch 3:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 9.79 | 7.48            | 2.31 
generator (test)       | 9.30 | 6.99            | 2.30 
discriminator (train)  | 1.20 | 0.00            | 1.20 
discriminator (test)   | 1.18 | 0.00            | 1.18 
Epoch 4 of 50
599/600 [============================>.] - ETA: 0s  
Testing for epoch 4:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 10.14 | 7.83            | 2.31 
generator (test)       | 9.91 | 7.61            | 2.30 
discriminator (train)  | 1.19 | 0.00            | 1.19 
discriminator (test)   | 1.17 | 0.00            | 1.17 
Epoch 5 of 50
599/600 [============================>.] - ETA: 0s  
Testing for epoch 5:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 10.27 | 7.96            | 2.31 
generator (test)       | 10.45 | 8.14            | 2.30 
discriminator (train)  | 1.19 | 0.00            | 1.19 
discriminator (test)   | 1.17 | 0.00            | 1.17 
Epoch 6 of 50
599/600 [============================>.] - ETA: 0s  
Testing for epoch 6:
component              | loss | generation_loss | auxiliary_loss
-----------------------------------------------------------------
generator (train)      | 10.56 | 8.25            | 2.31 
generator (test)       | 9.99 | 7.69            | 2.30 
discriminator (train)  | 1.18 | 0.00            | 1.18 
discriminator (test)   | 1.18 | 0.00            | 1.18

**Comment by s-macke** [comment]
I get the following output: ![mnist acgan git animation](https://github.com/s-macke/Kerasimo/blob/master/images/mnist_acgan.gif) 
when I use my script, which should be almost identical to the example except the learning rate
[mnist_acgan source](https://github.com/s-macke/Kerasimo/blob/master/models/mnist_acgan.py)

- Keras 2.0.2
- tensorflow 1.0.0
- cuda 8.0.61
- numpy 1.12.1
- cudnn 5.1.10

I successfully run it on a GTX 1080 and on a GTX 960.

**Comment by CanoeFZH** [comment]
after updating learning rate to 0.00001, I got excpected result
